text/0035-emergency-restarts.md

```diff
@@ -0,0 +1,116 @@
+
+Note that this has some desired properties:
+- We can specify any change to the state at any height without breaking the continuity of the chain. The first block after the restart is just like any other regular block, which means that eg. chain synchronization after the restart will work just like before the restart.
+- If the last safe block before the restart was at height `h`, but there were also blocks `h+1...h+n` that were finalized, they will become invalid after an upgrade point is specified at height `h+1`. Since the upgrade point wasn't included in the chainspec before the restart, the global state of the old block `h+1` won't satisfy the validity condition after an upgrade is added and the network is restarted. This means that all the blocks finalized after `h` will be discarded after restarting with the new chainspec and the network will resume operation at height `h+1`.
+
+### Consensus
+
+On the consensus side, it is important for an era change to happen at the boundary where the emergency restart happened. One of the reasons is that the fix for the issue that caused the crash could involve changes to the consensus logic, which could render the state of the last era invalid. This means that the conditions defining a switch block have to change as well: a switch block will be a block that either contains the `EraEnd` structure, or that has height `h-1` if there is an upgrade point specified for height `h`.
+
+### The structure of an upgrade point
+
+Currently, an upgrade point supports defining some new config values. We need it to also be able to specify some transforms that should be applied to the global state. This could be realized in multiple ways:
+
+1. Specifying the modifications as direct key-value pairs to be written to the global state (potentially overwriting the values under existing keys).
+2. Specifying arbitrary WASM to be executed, which would result in the transforms to be applied.
+3. For some special use cases, specifying changes in other ways, eg. support for changes to the validator set to be specified using an `accounts.csv`-like file.
+
+In this CEP, we suggest a combination of options 2 and 3.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The main thing required is to extend the `components::chainspec_loader::chainspec::UpgradePoint` structure to make it possible to specify arbitrary modifications to the global state.
+
+The best way of doing that is left as an unresolved question. We suggest adding the following two fields:
+
+```rust
+struct UpgradePoint {
+    // ... other fields
+    /// the paths to arbitrary WASM to be executed, in the specified order
+    upgrade_wasm_paths: Vec<String>,
+    /// the path to an `accounts.csv`-like file specifying the new set of validators
```

Comment:
Only validators? So this won't overwrite people's balances? We will use balances as they were at block `h-1` plus the potential changes from the `upgrade_wasm_paths`? 

I think we should stick to the Genesis-like approach where `accounts.csv` contains both validators' stakes and people's balances. 

cc @henrytill any opinion on this?